### PR TITLE
Update docs

### PR DESCRIPTION
**Pull request by cgmillette**
Updating documents from:
- What is rocWMMA
- Installation
- Programmers guide
- README.md

### CODE REVIEW

**Comment by samjwu** [comment]
For some reason the updated RTD config that uses mambaforge with doxygen 1.9.8 fails to pull the API functions correctly, though the original does it fine

Setting WARN_IF_DOC_ERROR to NO will avoid the Doxygen warning when multiple params have the same name

File: docs/api-reference-guide.rst [hunk 1]
```diff
@@ -5,99 +5,161 @@
...
 
-where,
+* Input Type = Matrix A / B
+* Output Type = Matrix C / D
+* Compute Type = Math / accumulation type
 
-Input Type = Matrix A/B
+* i8 = 8-bit precision integer
+* f8 = 8-bit precision floating point
+* bf16 = 8-bit precision brain floating point
```
Comment (dlangbe) [inline docs/api-reference-guide.rst]: bf8

File: docs/installation.rst [hunk 2]
```diff
@@ -87,27 +118,50 @@ On Centos, use:
...
+    pip3 install -r sphinx/requirements.txt
 
-You only need (library) for calling rocWMMA from your code.
+    python3 -m sphinx -T -E -b latex -d _build/doctrees -D language=en . _build/latex
+
+    cd _build/latex
+
+    pdflatex rocwmma.tex
+
+Running the above commands generates ``rocwmma.pdf``.
```
Comment (dlangbe) [inline docs/installation.rst:152]: Should these doc build instructions be identical in the README?

File: docs/installation.rst [hunk 3]
```diff
@@ -221,57 +280,63 @@ Build library and tests
...
 
-    CC=hipcc CXX=hipcc cmake -B<build_dir> .
+    CC=/opt/rocm/bin/amdclang CXX=/opt/rocm/bin/amdclang++ cmake -B <build_dir> . -DROCWMMA_BUILD_TESTS=ON
 
 After configuration, build using:
 
 .. code-block:: bash
 
-    cmake --build <build_dir> -- -j
+    cmake --build <build_dir> -- -j<nproc>
```
Comment (dlangbe) [inline docs/installation.rst:314]: Duplicate cmake instruction as line 255. Should consider consolidating all cmake build instructions.

File: docs/installation.rst [hunk 4]
```diff
@@ -280,12 +345,174 @@ To build the library and tests with assembly code generation, run:
...
 
-    CC=hipcc CXX=hipcc cmake -B<build_dir> . -DROCWMMA_BUILD_ASSEMBLY=ON
+    CC=/opt/rocm/bin/amdclang CXX=/opt/rocm/bin/amdclang++ cmake -B <build_dir> . -DROCWMMA_BUILD_ASSEMBLY=ON -DROCWMMA_BUILD_TESTS=ON
 
 After configuration, build using:
 
 .. code-block:: bash
 
-    cmake --build <build_dir> -- -j
+    cmake --build <build_dir> -- -j<nproc>
```
Comment (dlangbe) [inline docs/installation.rst:369]: Same comment about consolidation.

File: docs/programmers-guide.rst [hunk 5]
```diff
@@ -9,92 +9,225 @@
...
+--------------------------------
+
+rocWMMA is developed with the ``C++17`` language standard. The library takes advantage of several meta-programming techniques that help to statically
+optimize code at compile time and generate more efficient GPU kernels. All of the code is contained within the ``rocwmma`` namespace. Due to the templated
+nature of the code, the rocWMMA API is a header-only library. The developer has the advantage of full visibility of the implementation, as well as the ability to
+integrate rocWMMA API calls directly within their own kernels. An important feature is that the integrated code has higher visibility
+to the compiler's optimization passes where there is higher potential to generate more efficient device code. Moreover, no expensive host-device transfers or
+external kernel invokations are imposed by the API.
+
+The programming model that is the most useful with the rocWMMA API is wavefront-centric, in that all function calls will be assumed involve the entire wavefront. In a more general
```
Comment (dlangbe) [inline docs/programmers-guide.rst]: *assumed to involve

File: docs/programmers-guide.rst [hunk 6]
```diff
@@ -9,92 +9,225 @@
...
+Wavefronts will abstract blocks of data into ``fragments``, which are designed to encapsulate meta-data properties about the blocks in different contexts, along with the data itself:
+
+- a general geometric shape (e.g. BlockM/N/K dimensions)
+- a context of the provenance of the data (e.g. ``matrix_a`` or ``matrix_b``)
+- a context of how the data was stored (e.g. row or column major)
+- a datatype (e.g. single or half-precision floating point)
+
+These basic traits are then used to decide things like how much storage is needed, and how rocWMMA will organize individual threads in a layout to fetch or store the data.
+``Fragments`` are powerful objects because they are versatile in configuring and representing data, and their meta-properties propagate easily via Argument
+Dependent Lookup (ADL) and other deduction techniques. Of major importance to rocWMMA, the user first configures the fragments correctly and rocWMMA conveniently takes
```
Comment (dlangbe) [inline docs/programmers-guide.rst]: Reword "Of major importance..." to:
In practice, the user must simply configure the fragments correctly for their problem and rocWMMA conveniently takes care of the rest.

File: docs/programmers-guide.rst [hunk 7]
```diff
@@ -9,92 +9,225 @@
...
+care of the rest. The rocWMMA API can eliminate a very large amount of boiler-plate code as it provides tools to decompose matrix multiply-accumulate based problems into
```
Comment (dlangbe) [inline docs/programmers-guide.rst]: The last sentence seems out of place. It should be placed before the introduction of  what fragments are.

File: docs/programmers-guide.rst [hunk 8]
```diff
@@ -9,92 +9,225 @@
...
+- Vector operations. This level of objects such as ``OpaqueLoad`` or ``OpaqueStore`` handle variable-sized vector inputs and marshall them into unrolled unit backend operations. They encompass thread-wise details of vector operations. These classes provide a consistent functional interface for input vectors of all sizes, independent of device architecture, whose details are handled at a lower level.
+- Fragment operations. This is the API level of rocWMMA where user data is stored and manipulated in ``fragment`` objects. Fragment objects can be visualized as geometric 'blocks' of data in the perspective of the current wavefront, which are to be stored as vectors. Each of the loading / storing and mma operations provided by rocWMMA are in the perspective of the wavefront, assuming that all threads in the wavefront are participating under the hood. This layer's implementation translates wavefront fragment operations into vector operations and so on. This way, the rocWMMA API experience intends to be seamless across different device architectures and platforms.
```
Comment (dlangbe) [inline docs/programmers-guide.rst]: Reword "This way, the rocWMMA API..." to:
The encapsulation of rocWMMA into these separate layers allows for an API experience that is seamless across different device architectures and platforms.

File: docs/programmers-guide.rst [hunk 9]
```diff
@@ -9,92 +9,225 @@
...
+    MB - Multiple output blocks targeted per wave
+    SB - Single output block target per wave
+    NC - Non-Cooperative load / store
+    CP - Cooperative load / store
+    BLK - Cooperative load / store per block tile
+    WV - Cooperative load / store per wave tile
+    WG - Cooperative load / store per macro tile
+
+* ``gemm_PGR0_LB0_MP0_SB_NC``: The simplest blocked GEMM example, which targets one output
+  block of matrix multiplication per wave. No prefetch, no lDs usage, default MFMA prioritization, single
```
Comment (dlangbe) [inline docs/programmers-guide.rst]: Should this and the following occurrences of "lDs" be capitalized like this?

File: docs/what-is-rocwmma.rst [hunk 10]
```diff
@@ -1,25 +1,32 @@
...
-rocWMMA where WMMA stands for Wavefront Mixed precision Multiply Accumulate, is AMD's C++ library for accelerating mixed precision matrix multiply-accumulate operations
-leveraging specialized GPU matrix cores on AMD's latest discrete GPUs.
+rocWMMA is a C++ header library for accelerating mixed precision matrix multiply-accumulate operations
+leveraging specialized GPU matrix cores on AMD's latest discrete GPUs. 'roc' being an AMD-specific
+component belonging to the ROCm ecosystem, and WMMA stands for Wavefront Mixed precision Multiply Accumulate.
 
-The C++ APIs facilitate the decomposition of matrix multiply-accumulate problems into
-discretized block fragments and parallelize the block-wise operations across multiple GPU wavefronts.
+rocWMMA leverages modern C++ techniques. It is templated for modularity and uses meta-programming paradigms to provide opportunities for customization
+and compile-time inferences and optimizations. The API is seamless across supported CDNA and RDNA architectures. It is also portable with Nvidia
```
Comment (dlangbe) [inline docs/what-is-rocwmma.rst]: *portable with the...

File: docs/what-is-rocwmma.rst [hunk 11]
```diff
@@ -1,25 +1,32 @@
...
+Major benefits include kernel-level control which allows authoring flexibility and accessibility to compiler optimization passes in-situ
+with other device code. Users can therefore decide when and where kernel run-time launches are required, which is not dictated by the API.
 
-As rocWMMA is written in C++, it can be applied directly in the device kernel code. The library code is templated for modularity and uses the available meta-data to provide opportunities for compile-time inferences and optimizations.
+rocWMMA's API facilitates the decomposition of matrix multiply-accumulate problems into discretized blocks (also known as fragments) and enables
+parallelization of block-wise operations across multiple GPU wavefronts. The programmer's perspective is simplified to wavefront handling of fragments,
+whereas individual threads are handled internally. This can allow for faster development times and a more seamless experience across multiple architectures.
+API functions consist data loading and storing, matrix multiply-accumulate and helper transforms operating on data fragment abstractions. Moreover, data movement
```
Comment (dlangbe) [inline docs/what-is-rocwmma.rst]: Consider changing to:
API functions include data loading and storing, matrix multiply-accumulate and helper transforms that operate on data fragment abstractions.

File: docs/programmers-guide.rst [hunk 12]
```diff
@@ -9,92 +9,225 @@
...
-  - Packing and unpacking
-  - Conversion and broadcasting
-  - Load and store
+  - Loading and storing utilities
+  - Layouts of memory and registers
+  - Mapping utilities
+  - Intrinsic wrappers
+  - Vector class implementations
+  - Vector conversion, permutation and transform utilities
+  - Vector packing and unpacking
```
Comment (mkarunan) [inline docs/programmers-guide.rst:185]: - data layout transformation utilities

File: docs/installation.rst [hunk 13]
```diff
@@ -57,24 +69,43 @@ packages as described above. If still desired, here are the instructions to buil
...
+.. note::
+    \* = if using rocBLAS for validation.
+
+    It is best to use available ROCm packages from the same release where applicable.
 
 Download rocWMMA
 ^^^^^^^^^^^^^^^^^
 
-The rocWMMA source code is available at the `rocWMMA github page <https://github.com/ROCmSoftwarePlatform/rocWMMA>`_. rocWMMA has a minimum ROCm support version 5.4.
+The rocWMMA source code is available at the `rocWMMA github page <https://github.com/ROCmSoftwarePlatform/rocWMMA>`_. rocWMMA has a minimum ROCm support version 6.0.
```
Comment (CongMa13) [inline docs/installation.rst]: This link is still alive. But wmma has moved to https://github.com/ROCm/rocWMMA

File: library/include/rocwmma/rocwmma.hpp [hunk 14]
```diff
@@ -34,184 +34,119 @@
...
- * WaveSize*4 |   1       |
- *
- * @note TBlockX must be a multiple of WaveSize
- *
- *
- * **Supported Matrix Layouts**
+ * rocWMMA is a C++ header library for accelerating mixed precision matrix multiply-accumulate operations leveraging specialized GPU matrix cores on AMD's latest discrete GPUs.
+ * 'roc' being an AMD-specific component belonging to the ROCm ecosystem, and WMMA stands for Wavefront Mixed precision Multiply Accumulate.
+ * rocWMMA leverages modern C++ techniques. It is templated for modularity and uses meta-programming paradigms to provide opportunities for customization and compile-time inferences
+ * and optimizations. The API is seamless across supported CDNA and RDNA architectures. It is also portable with Nvidia nvcuda::wmma library, allowing those users to easily migrate
```
Comment (dlangbe) [inline library/include/rocwmma/rocwmma.hpp]: Consider the same requested changes as in what-is-rocwmma.rst

File: docs/installation.rst [hunk 15]
```diff
@@ -221,57 +280,63 @@ Build library and tests
...
+``unit/load_store_matrix_coop_sync_test``       Tests ``load_matrix_coop_sync`` and ``store_matrix_coop_sync`` API functions
+``unit/map_util_test``                          Tests mapping utilities used in rocWMMA implementations
+``unit/pack_util_test``                         Tests vector packing utilities used in rocWMMA implementations
+``unit/transforms_test``                        Tests transform utilities used in rocWMMA implementations
+``unit/unpack_util_test``                       Tests vector un-packing utilities used in rocWMMA implementations
+``unit/vector_iterator_test``                   Tests internal vector storage iteration implementation
+``unit/vector_test``                            Tests internal vector storage implementation
+``unit/vector_util_test``                       Tests internal vector manipulation utilities implementation
+============================================= ===================================================================================================================================================
+
```
Comment (CongMa13) [inline docs/installation.rst:349]: render result of these lines:
image

expect a Note

image

**Review by CongMa13 — CHANGES_REQUESTED** [review]
The new document looks awesome! Especially the programmer's guide.

File: docs/installation.rst (hunk 2 above)
Comment (cgmillette) [reply docs/installation.rst:152]: Added both html and pdf instructions to both areas

File: docs/installation.rst (hunk 3 above)
Comment (cgmillette) [reply docs/installation.rst:314]: The initial cmake call is for the configuration

Subsequent calls to cmake --build are invoking the compiler, which should be now configured.

**Comment by cgmillette** [comment]
Feedback has been incorporated

**Review by mkarunan — APPROVED** [review]
Looks great!

**Review by CongMa13 — APPROVED** [review]

**Pull request merged**